server/service/src/programs/contact_trace/upsert.rs

```diff
@@ -0,0 +1,501 @@
+        })
+        .map_err(|err: TransactionError<UpsertContactTraceError>| err.to_inner_error())?;
+    Ok(document)
+}
+
+impl From<RepositoryError> for UpsertContactTraceError {
+    fn from(err: RepositoryError) -> Self {
+        UpsertContactTraceError::DatabaseError(err)
+    }
+}
+
+fn generate(
+    user_id: &str,
+    input: UpsertContactTrace,
+    registry: DocumentRegistry,
+    existing: Option<Document>,
+) -> Result<RawDocument, RepositoryError> {
+    let now = Utc::now();
+    Ok(RawDocument {
+        name: existing.map(|e| e.name).unwrap_or(patient_doc_name_with_id(
+            &input.patient_id,
+            &input.r#type,
+            &now.to_rfc3339(),
+        )),
+        parents: input.parent.map(|p| vec![p]).unwrap_or(vec![]),
+        author: user_id.to_string(),
+        datetime: now,
+        r#type: input.r#type.clone(),
+        data: input.data,
+        form_schema_id: Some(input.schema_id),
+        status: DocumentStatus::Active,
+        owner_name_id: Some(input.patient_id),
+        context_id: registry.context_id,
+    })
+}
+
+fn validate_contact_trace_schema(
+    input: &UpsertContactTrace,
+) -> Result<SchemaContactTrace, serde_json::Error> {
+    // Check that we can parse the data into a default Program object, i.e. that it's following the
+    // default program JSON schema.
+    // If the program data uses a derived program schema, the derived schema is validated in the
+    // document service.
+    serde_json::from_value(input.data.clone())
+}
+
+fn validate_patient_exists(
+    ctx: &ServiceContext,
+    patient_id: &str,
+) -> Result<bool, RepositoryError> {
+    let patient = PatientRepository::new(&ctx.connection)
+        .query_by_filter(
+            PatientFilter::new().id(EqualFilter::equal_to(patient_id)),
+            None,
+        )?
+        .pop();
+    Ok(patient.is_some())
+}
+
+fn validate_document_type(
+    ctx: &ServiceContext,
+    document_type: &str,
+) -> Result<Option<DocumentRegistry>, RepositoryError> {
+    let mut entry = DocumentRegistryRepository::new(&ctx.connection).query_by_filter(
+        DocumentRegistryFilter::new()
+            .r#type(DocumentRegistryCategory::ContactTrace.equal_to())
+            .document_type(EqualFilter::equal_to(document_type)),
+    )?;
+    Ok(entry.pop())
+}
+
+fn validate_program(
+    ctx: &ServiceContext,
+    context_id: &str,
+) -> Result<Option<ProgramRow>, RepositoryError> {
+    ProgramRepository::new(&ctx.connection)
+        .query_one(ProgramFilter::new().context_id(EqualFilter::equal_to(context_id)))
+}
+
+fn validate_parent_doc_exists(
+    ctx: &ServiceContext,
+    parent: &str,
+) -> Result<Option<Document>, RepositoryError> {
+    DocumentRepository::new(&ctx.connection).find_one_by_id(parent)
+}
+
+struct ValidationResult {
+    contact_trace_data: SchemaContactTrace,
+    document_registry: DocumentRegistry,
+    program_row: ProgramRow,
+    parent: Option<Document>,
+}
+
+fn validate(
```

Comment:
Partially aggree, was just to merge some of those but then stopped because there are subtle differences, sometimes requiring a new parameter... I am tending to leave as it for now as they don't really have major bussines logic in them, mostly one liners.

What I like right now is that everything is in one file and you can see on one page what the validations do without hoping around.

Reply:
> I am tending to leave as it for now as they don't really have major business logic in them, mostly one liners.
> What I like right now is that everything is in one file and you can see on one page what the validations do without hoping around.

I can see where you coming from, although would still need to hop around the file. I am just worried that if pattern is not created from the start, it wont' be created later (i.e. if there is a change to any of the simple one liner validations in the future, i doubt they will be consolidated, in fact only some may change as it would be hard to find all of the validations). Also the idea for the validation method helpers was to make validators easier to read, and give reader the choice of only going deeper if they need to (so most of the time you wouldn't need to hop around to the underlying implementation). Lastly, for reviews, if common methods are used, it's less code to review (reviewer can trust that existing code is correct).

Reply:
Personally, I found it to be the better strategy to delay such consolidations till there is good reason for it. For example, for very simple methods like most validation methods I find it more straight forward to just inline them, IMHO it then becomes easier to read and review ;-). For example, `validate_document_type()` actually does something in the line of `validate_document_type_and_get_matching_document_registry()` ... But yes there is the danger that something changes unexpectedly or that there is a bug, and then it's less clear where it needs to be fixed as well.
 
Not strong about it but maybe low priority for this PR?

Reply:
Yeah, i am not super strong about it, and not sure if my concern is validated, I just wanted to explain why i mentioned it. I guess we just need to be mindful of the fact that similar validation logic exist in different places, and if business logic ever changes (probably not likely so not important) we should update in multiple places